cmd/traffic/cmd/manager/internal/state/intercept.go

```diff
@@ -410,3 +410,62 @@ func findIntercept(ac *agentconfig.Sidecar, spec *managerrpc.InterceptSpec) (fou
 	}
 	return nil, nil, errcat.User.Newf("%s %s.%s has no interceptable port%s", ac.WorkloadKind, ac.WorkloadName, ac.Namespace, ss)
 }
+
+type InterceptFinalizer func(ctx context.Context, interceptInfo *managerrpc.InterceptInfo) error
+type interceptRefresher func() (*managerrpc.InterceptInfo, bool)
+
+type interceptState struct {
+	sync.Mutex
+	lastInfoCh       chan *managerrpc.InterceptInfo
+	finalizers       []InterceptFinalizer
+	refreshIntercept interceptRefresher
+	interceptID      string
+}
+
+func newInterceptState(clientCtx context.Context, tmCtx context.Context, interceptID string, refreshIntercept interceptRefresher) *interceptState {
+	is := &interceptState{
+		lastInfoCh:       make(chan *managerrpc.InterceptInfo),
+		interceptID:      interceptID,
+		refreshIntercept: refreshIntercept,
+	}
+	go is.monitor(clientCtx, tmCtx)
+	return is
+}
+
+func (is *interceptState) addFinalizer(finalizer InterceptFinalizer) {
+	is.Lock()
+	defer is.Unlock()
+	is.finalizers = append(is.finalizers, finalizer)
+}
+
+func (is *interceptState) terminate(info *managerrpc.InterceptInfo) {
+	is.lastInfoCh <- info
+}
+
+func (is *interceptState) monitor(clientCtx context.Context, tmCtx context.Context) {
+	for {
```

Comment:
Why does this happen in a loop? Seems all cases in the `select` will return.

Reply:
great catch that's what I get for iterating so much over the code

Reply:
This code is actually gone now, pls check [this commit](https://github.com/telepresenceio/telepresence/pull/2594/commits/f33f4307c51bf84a9aee4b27d99609c27876a060) because I replaced it with a simpler solution